### PR TITLE
mpi: Fix data distribution bugs [part 1]

### PR DESCRIPTION
**Pull request by rhodrin**
This set of PRs will supersede #1888.

Later PRs will deal specifically with issues #1010 and #1862.

This PR fixes some MPI data slicing issues. See the newly added [tests](https://github.com/devitocodes/devito/blob/62a9574e6bfb0d65fa817c696b3f8292e40c2b99/tests/test_data_parallel.py#L8) for details.

Note that these tests are located in a new `test_data_parallel.py` file to avoid `test_data.py` becoming too unwieldy. We may wish to refactor these two files in a little in this or an upcoming PR?

### CODE REVIEW

File: devito/data/data.py [hunk 1]
```diff
@@ -247,7 +247,12 @@ def __getitem__(self, glb_idx, comm_type, gather_rank=None):
                 else:
                     pass
                 it.iternext()
-            return retval
+            reshape = tuple([s for s, i in zip(retval.shape, loc_idx)
+                             if type(i) is not np.int64])
+            if reshape and 0 not in reshape and reshape != retval.shape:
```
Comment (mloubout) [inline devito/data/data.py:252]: maybe some parenthesis for readability?

File: devito/data/utils.py [hunk 2]
```diff
@@ -389,7 +389,7 @@ def flip_idx(idx, decomposition):
                 stop = i.stop
             processed.append(slice(start, stop, i.step))
         else:
-            processed.append(i)
+            processed.append(slice(i, i+1, 1))
```
Comment (mloubout) [inline devito/data/utils.py:392]: `1` or `None`?

File: tests/test_data_parallel.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import pytest
+import numpy as np
+
+from devito import Grid, Function
+from devito.data import loc_data_idx
+
+
+class TestDataParallel(object):
```
Comment (mloubout) [inline tests/test_data_parallel.py:8]: Can maybe be moved in `test_mpi`?

**Review by mloubout — COMMENTED** [review]
Sweeet thanks @rhodrin for finding the root of it

File: devito/data/data.py [hunk 4]
```diff
@@ -247,7 +247,12 @@ def __getitem__(self, glb_idx, comm_type, gather_rank=None):
                 else:
                     pass
                 it.iternext()
-            return retval
+            reshape = tuple([s for s, i in zip(retval.shape, loc_idx)
```
Comment (FabioLuporini) [inline devito/data/data.py:250]: I would add a comment about what's being done here

**Comment by FabioLuporini** [comment]
IMHO we either move the test inside here: https://github.com/devitocodes/devito/blob/master/tests/test_data.py#L475
or move the class inside test_data_parallel

other than that, uncontroversial (perhaps add a couple of comments to the new code as per my comment above)

File: devito/data/data.py (hunk 1 above)
Comment (rhodrin) [reply devito/data/data.py]: Ta, added those.

File: devito/data/utils.py (hunk 2 above)
Comment (rhodrin) [reply devito/data/utils.py:392]: Doesn't matter I don't think? `1` is just more 'explicit'?

File: tests/test_data_parallel.py (hunk 3 above)
Comment (rhodrin) [reply tests/test_data_parallel.py]: I've moved them to `TestDistributedData` in `test_data.py` now.

**Comment by rhodrin** [comment]
> Not sure I like a whole new file for just one single test

Yeah, move back to `test_data.py` for now. (Although note that the reason a new file was added for this PR was that many more tests will be added in the next couple of PRs).

File: tests/test_data.py [hunk 5]
```diff
@@ -1273,6 +1273,56 @@ def test_misc_data(self):
             assert(np.all(c.data[-2:-1] == [3., 3., 5., 8., 7.]))
             assert(np.all(c.data[-1] == [4., 4., 6., 9., 8.]))
 
+    @pytest.mark.parallel(mode=4)
```
Comment (FabioLuporini) [inline tests/test_data.py:1276]: perhaps `mode=[4, 8]` is it worth it or overkill?
or perhaps `mode=[4, 7]` 😬

**Pull request merged**